[JIT] Add EliminateExceptions pass.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54732 [JIT] Add EliminateExceptions and RemoveInplaceOps to PE pass pipeline.
* **#54730 [JIT] Add EliminateExceptions pass.**

Differential Revision: [D27343165](https://our.internmc.facebook.com/intern/diff/D27343165)

Commit:
[JIT] Add EliminateExceptions pass.

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/remove_exceptions.cpp:
```diff
@@ -0,0 +1,49 @@
+
+bool certainlyThrows(Block* block) {
+  for (Node* n : block->nodes()) {
+    if (n->kind() == prim::RaiseException) {
```
I'm not quite grokking the big picture here, but I would guess it will be used in cases where a block would contain *only* `raise Exception` and *no* statements capable of altering the global state?

ZolotukhinM replied:
The assumption here is that the model doesn't throw exceptions and thus `prim::RaiseException` can be used as an indicator of a never executed block. I will add a comment covering that, and hopefully it would address your other concern too by making the purpose of the pass more clear.

Krovatkin commented on torch/csrc/jit/passes/remove_exceptions.cpp:
```diff
@@ -0,0 +1,49 @@
+}
+
+void EliminateExceptions(std::shared_ptr<Graph>& graph) {
+  GRAPH_DUMP("Before EliminateExceptions: ", graph);
```
I would stick a check here in case someone decides that it's a useful-sounding pass to be called unconditionally O_o

Krovatkin reviewed (approved):
Would be great if you could add some context to the PR description

Commit:
Update on "[JIT] Add EliminateExceptions pass."


Differential Revision: [D27343165](https://our.internmc.facebook.com/intern/diff/D27343165)

[ghstack-poisoned]
